lib/web_ui/lib/painting.dart

```diff
@@ -491,6 +491,42 @@ Future<Codec> instantiateImageCodecFromBuffer(
   targetHeight: targetHeight,
   allowUpscaling: allowUpscaling);
 
+Future<Codec> instantiateImageCodecWithSize(
+  ImmutableBuffer buffer, {
+  TargetImageSizeCallback? getTargetSize,
+}) async {
+  if (getTargetSize == null) {
+    return engine.renderer.instantiateImageCodec(buffer._list!);
+  } else {
+    final Codec codec = await engine.renderer.instantiateImageCodec(buffer._list!);
+    try {
+      final FrameInfo info = await codec.getNextFrame();
+      try {
+        final int width = info.image.width;
+        final int height = info.image.height;
+        final TargetImageSize targetSize = getTargetSize(width, height);
+        return engine.renderer.instantiateImageCodec(buffer._list!,
+            targetWidth: targetSize.width, targetHeight: targetSize.height, allowUpscaling: false);
+      } finally {
+        info.image.dispose();
+      }
+    } finally {
+      codec.dispose();
+    }
+  }
+}
+
+typedef TargetImageSizeCallback = TargetImageSize Function(int intrinsicWidth, int intrinsicHeight);
+
+class TargetImageSize {
```

Comment:
Are there other places in dart:ui we could use an int-based size? Might make sense to generalise the name so others could use it later, if so.

I can’t think of a great name though. Maybe something like `PhysicalSize`.

Reply:
I thought about that, but a real `PhysicalSize` class would want both width & height to be non-null.  This is unique in how it allows one or both of those to be null.

Reply:
More similar to an IntConstraints type, not quite a size. We do really need an ISize type, but probably unrelated.